### ISSUE TITLE
Couchbase sync running with Couchbase cluster with auto failover is trying to connect to node that went down.

### ISSUE THREAD

**Issue by krzysztofkolski** [couchbase/sync_gateway#2425]
Hi,

We have a strange problem with Couchbase Sync (1.4 Community) that is connected to Couchbase Cluster (4.5 Community) with autofailover option turned on.

Few information about our configuration (with IP's so they can be matched in logs).
- Three nodes  of Couchbase Cluster (with autofailover option turned on (after 30 seconds): available at 10.13.0.4,10.13.0.5,10.13.0.6 (3 separated Windows 2012 machines)
- Two instances of Couchbase Sync (available at 10.14.0.4, 10.14.0.5) (two instances of Windows 2012 machines) (in configuration they are pointing to server: http://10.13.0.4:8091)
- Couchbase sync and Couchbase server nodes are on separate machines.

### Steps to reproduce the problem
1. Start both sync gateways
2. Stop service CB at 10.13.0.4 -> action properly handled by couchbase cluster autofailover after configured time of 30s.
3. Couchbase sync's were reconnected to 10.13.0.6 ( as can be seen in logs)
4. Start Couchbase service at 10.13.0.4 (the same as was stopped in step 2), make delta recovery and rebalancing -> all 3 nodes of CB cluster are now up and running.
5. Stop Couchbase node service at 10.13.0.6.
After step 5 the couchbase sync should reconnect to one of two available nodes 10.13.0.4 or 10.13.0.5 but as can be seen in logs it tries to connect to 10.13.0.6 - that was turned off. After that operation couchbase sync db went to offline state and only option to repair it from this state is to restart couchbase sync process (or call API method) because after Couchbase Server node 10.13.0.6 was started again Sync Gateway DB didn't recover from it's offline state automatically

Why Couchbase Sync is trying to connect to Couchbase Server node that is down. It should - as in first time try to connect to one of two nodes that are up and running (in this case 13.*.04 and 13.*.05. As can be seen in logs, after step 5 -where cb node was turned off - couchbase sync received new configuration (2017-03-27T15:37:26.904+00:00 [INFO] Got new configuration for bucket catalogueData) so it should know that 13.*.06 is not available. 
How Couchbase Sync is selecting node to which it tries to connect - it's a random choice?

I will be grateful for your help with , since now after windows restart (for example after update) we need to manually restart both couchbase sync processes to be 100% sure they are working correctly. 

 
### Sync Gateway version
Couchbase Sync Gateway/1.4.0(2;9e18d3e)
### Operating system
Windows 2012

### Log output
Links to logs from both sync gateways
https://gist.github.com/krzysztofkolski/008c80837939083fdd8ec811eac6ae6f
https://gist.github.com/krzysztofkolski/27b3e073b0b5501ff09bb828770e3cc0
### Expected behavior
Sync gateways should know which node in couchbase cluster is up and running and should try to connect to one of them, instead of trying to connect to the one that just went off, and making db offline.

### Actual behavior
Currently it happens (irregulary), that couchbase sync is trying to connect to couchbase node that was just turned off even when autofailover at couchbase server was correct.

**Comment by tleyden** [comment]
@krzysztofkolski in the Sync Gateway configuration, which Couchbase node is it pointing to?

**Comment by tleyden** [comment]
@krzysztofkolski actually n/m, I can see from the logs that Sync Gateway is configured to connect to `10.13.0.6`

**Comment by tleyden** [comment]
> but as can be seen in logs it tries to connect to 10.13.0.6

In this case Sync Gateway is creating a brand new connection to Couchbase Server, so when the database comes back online it will look into the config to determine which IP address to connect to.

The fix that will address the root of the problem is to allow the Sync Gateway configuration to take an array of Couchbase Server URL's in it's config: https://github.com/couchbase/sync_gateway/issues/1019, so I'll close this as a duplicate.

In the meantime, the best workaround I can think of is to use [DNS SRV records](https://developer.couchbase.com/documentation/server/current/sdk/java/managing-connections.html#story-h2-6)  Aside from that doc there is also there is a [blog post](http://nitschinger.at/Bootstrapping-from-DNS-SRV-records-in-Java/).

@ajres if it sounds like it was a requirement for the db online/offline feature to keep a history of couchbase server nodes that it previously connected to, rather than just relying on the Sync Gateway configuration, can you re-open this ticket?  I'm assuming it wasn't.

**Comment by krzysztofkolski** [comment]
SyncGateway in its configuration is configured to use 10.13.0.4 server node. When 10.13.0.4 goes down (and autofailover happens), it reconnects automatically to 10.13.0.6 (which is ok). Then 10.13.0.4 goes up, and 10.13.0.6 went down. And here is a problem. Couchbase Sync is trying to connect one more time to 10.13.0.6 that just went down and then moving db to offline state, instead of trying to 10.13.0.4 or 10.13.0.5. 

I'll look into this DNS SRV records you mentioned. 

When you plan to implement issue #1019 you mentioned above?

**Comment by krzysztofkolski** [comment]
@tleyden
In documents you pointed to me there is described how to this  magic with DNS SRV records where I'm using our own code to connect to couchbase cluster (in example from docs in Java). 
In our case we have couchbase sync that is connecting to couchbase cluster - so no custom code where we can configure usage of dns srv record as it was desribed in links you provided.

Can you point me to some solution that describes how to configure usage if this dns srv records in couchbase sync configuration? 
  
To summarize. The question remains open, how to configure high availibility with couchbase cluster and couchbase sync - since like I described above at some point sync is trying to connect to the node that just went down)

Please look that initially when Sync was started it connects to 10.13.0.4 (which was provided in configuration). When this node went down, Sync automatically was reconnected to the next node (10.13.0.6) - so in this first case the behaviour was correct. And then I started couchbase node at 10.13.0.4 (that was previosly killed) so now I have proper 3 node working couchbase cluster. Then I'm killing 10.13.0.6 - so the node to which sync was reconected after first failover, and then the problem happens - sync should reconnect to 10.13.0.4 or 10.13.0.5 so to nodes that are available on cluster but it tries to 10.13.0.6 that was killed and then db is going offline.

**Comment by tleyden** [comment]
> Can you point me to some solution that describes how to configure usage if this dns srv records in couchbase sync configuration?

Here's how I imagine it would work:

1. You need to create a DNS SRV record like couchbase.yourcompany.com
1. For each Couchbase Server node, you will want to add their IP address to the DNS SRV record.  This could be done manually or as part of a script.
1. Add some monitoring to detect failed Couchbase Server Nodes and *remove* them from the list of available servers in the DNS SRV record.  
1. You would re-configure your Sync Gateway to point to `couchbase.yourcompany.com:8091`, and then it would route to one of the active Couchbase Server Nodes.

@ingenthr / @daschl does that sound about right?

**Comment by krzysztofkolski** [comment]
Welcome

Considering Sync Gateway as a client using Couchbase HA cluster proposed solution is rather a quirk.

Sync Gateway has serious flaw in bootstrap procedure (https://github.com/couchbase/sync_gateway/issues/1019).

Using DNS SRV as proposed below (“manual” high availability ?!?!) is surprising considering SDK published here https://developer.couchbase.com/documentation/server/current/sdk/java/managing-connections.html#story-h2-6 explaining DNS SRV record usage procedure available for custom code connecting directly to Couchbase.

SDK is not compliant as well with RFC 2782 (https://www.ietf.org/rfc/rfc2782.txt https://tools.ietf.org/html/rfc2782 ) but it is a small improvement.

Why Sync Gateway cannot  actually follow quite simple usage rules as described in RFC2782 (published February 2016) is a mystery.
Please consult “Overview and rationale” RFC 2782 (https://www.ietf.org/rfc/rfc2782.txt https://tools.ietf.org/html/rfc2782 ) to understood sole purpose of SRV records existence and difference from A records.
Please consult “Usage rules” starting on page 5 RFC 2782 (https://www.ietf.org/rfc/rfc2782.txt https://tools.ietf.org/html/rfc2782 ) to understood why proposed manual/external monitoring steps are preposterous.

Excerpt from RFC 2782:
“
Sort the list by priority (lowest number first)

            Create a new empty list

            For each distinct priority level
                While there are still elements left at this priority
                level

                    Select an element as specified above, in the
                    description of Weight in "The format of the SRV
                    RR" Section, and move it to the tail of the new
                    list

            For each element in the new list

                query the DNS for address records for the Target or
                use any such records found in the Additional Data
                section of the earlier SRV response.

                for each address record found, try to connect to the
               (protocol, address, service).
“

While there are still other considerations (e.g. proper discovery using SRV of one from many Couchbase clusters in same domain can be solved with help of https://tools.ietf.org/html/rfc6763) this simple recipe allow to detect and test each Couchbase node during Sync Gateway bootstrap to obtain current cluster configuration.

Understanding proper DNS SRV record usage rules will render https://github.com/couchbase/sync_gateway/issues/1019 obsolete since all Couchbase nodes configured in SRV record will be tried in proper order as priority and weight defines.

Even simple stupid routine “try all in any order from SRV list” will solve this problem if implemented in Sync Gateway and SDK CouchbaseCluster.create().

I admire Traun way of problem slowing by closing BOTH requests https://github.com/couchbase/sync_gateway/issues/2070 and https://github.com/couchbase/sync_gateway/issues/2058 as duplicate of each other when one those was error while the other (2070) is functional change that is actually way different from issue 1019 and can be useful in advanced networking scenarios.

Please be advised about existence of  “Internet Assigned Numbers Authority (IANA) Procedures for the Management of the Service Name and Transport Protocol Port Number Registry” https://tools.ietf.org/html/rfc6335 and http://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xml
Arbitrary creation of names like “_couchbase._tcp” or “_couchbases._tcp” by organization that is “part of the $35 billion database market and has quickly built a reputation among the Fortune 500” seems to be negligent.

**Comment by tleyden** [comment]
Hey @krzysztofkolski -- I get that you are frustrated, I know the feeling .. there is a lot of tech that drives me nuts too, BUT I want to let you know that this post doesn't meet our code of conduct.  

To be fair, we actually haven't published a code of conduct yet, which will change soon .. but in the meantime, please refer to the [Golang community code of conduct](https://golang.org/conduct).  Going forward for any github posts, everyone will be expected to follow this -- both users and Couchbase employees.  

If you want to know what I'm referring to specifically, my email address is on my github profile and feel free to send me an email.  I know I speak for the whole team when I say we want to create a safe place to communicate in the open.  

I'll send follow up responses to your specific points in a separate post.

**Comment by tleyden** [comment]
Regarding DNS SRV vs a list of hosts -- I don't think it's a "one or the other" scenario.  As stated in the [docs regarding DNS SRV](https://developer.couchbase.com/documentation/server/current/sdk/java/managing-connections.html#story-h2-6):

> As an alternative to specifying multiple hosts in your program, you can get the actual bootstrap node list from a DNS SRV record

In other words, users need to be able to have either approach available to them, and specifying multiple hosts (#1019) will never become obsolete.  You could argue that DNS SRV is a better approach, but the point I'm trying to make is that the ability to specify multiple hosts is a core feature that will never change.

So far in the Sync Gateway product requirements, DNS SRV has not come up yet to my knowledge.  I threw it out as a workaround until #1019 is completed.  But I definitely think that DNS SRV support merits first class support.  From what I can tell, it wouldn't require any changes in Sync Gateway itself or the SDK's, it's really up to the user to correctly configure their DNS SRV record.  

> closing BOTH requests #2070 and #2058 as duplicate of each other when one those was error while the other (2070) is functional change that is actually way different from issue 1019 and can be useful in advanced networking scenarios.

#2070 and #2058 were both closed as duplicates of #1019, not as each other.  How do you see #2070 as being any different from #1019?  

> SDK is not compliant as well with RFC 2782 (https://www.ietf.org/rfc/rfc2782.txt https://tools.ietf.org/html/rfc2782 ) but it is a small improvement.

Would you mind [filing an issue](https://issues.couchbase.com/secure/Dashboard.jspa) against the SDK with details?

**Comment by ingenthr** [comment]
Note @krzysztofkolski that I was the instigator for use of SRV records.  Is the concern on compliance that SDKs do not follow weighting or priority as recommended in IETF RFC 2782?  The best place probably for an issue on this would be [in the sdk-rfc repository](https://github.com/couchbaselabs/sdk-rfcs).  The original writeup on our use of SRV records predates the sdk-rfc repo, but that'd be the logical place for us to update/track issues cross SDK.

**Comment by krzysztofkolski** [comment]
Hej @tleyden, sorry for my harsh reposne. Yes you're right that it is very frustrating that the option of having high avalilibility platform with couchbase and couchbase sync is not working well. Additionally your workaround proposition where I need to change SRV records dynamically by some external script is not aplicable in my case, because not all domain providers allows to do this. Maybe you have some other workaround solution that you can propose? The only I can think of right now is to write some additional script/software that will be triggered by scheduler every 10-15 minutes and will be checking if couchbase sync db is still responding and online - and if not - it will restart couchbase sync process - but to be honest this is not very profesional aproach on production environment.
